core: Adjust manager startup time from monotonic time

filbranden opened the pull request:
When the RTC time at boot is off in the future by a few days, systemd will record and report that time as manager startup time in `systemd-analyze dump` and `"UserspaceTimestamp"` D-Bus properties.

Update the logic so that the time will be adjusted based on monotonic time. We do that by calculating the adjusted manager startup realtime from the monotonic time stored at that time, by comparing that time with the realtime and monotonic time of the current time.

This is a follow up to PR #16428 and related to Issue #16347. (cc @anitazha)

I'm not really sure whether this is the best solution... If the realtime clock for some reason fluctuates and starts becoming unreliable, then the manager startup time might start to drift, which might just look odd... So I think we might need more discussion on this one before merging. Maybe it would make most sense to do a single one-time correction whenever `time-sync.target` is started, assuming it's ever pulled? Any opinions either way?

Thanks!
Filipe

poettering commented:
Hmm, the idea was to record the actual timestamps in effect then, so that we can relate them with log timestamps and such. You break this that way.

What's the precise rationale for this? i.e. where do you see these timestamps where you wished you would see the corrected timestamps instead? isn't this something the display code should handle? i.e. record the raw data and fix up the wallclock timestamps only on display?

filbranden commented:
This came up in #16347, where Anita was using `systemd-analyze dump | grep userspace` to show that this clock was set in the future.

Yeah I agree this one is questionable, that's why I split it from the other one with the fix to the timer.

I guess the intent here is to solve the following problem: If I query the time when systemd came up and it tells me it was one day in the future, is that confusing or helpful? Or, if it tells me it started 15 minutes ago but I know it's been up for days, is that more confusing or helpful?

Or, more generally, should we display the timestamps we record at points where there's a high enough probability that they might be incorrect? Should we flag them as such to the user? When we do have reliable timestamps around (`time-sync.target` has been reached) should we take the opportunity to adjust them, so they match reality? Should we maybe display both recorded and adjusted?

Not a ultra high priority issue, just wondering if we can be more helpful and friendly here in such cases.

(I'll go ahead and close this one, I don't think we'll end up using it.)
